crates/oxc_linter/src/rules/unicorn/numeric_separators_style.rs

```diff
@@ -0,0 +1,667 @@
+use lazy_static::lazy_static;
+use oxc_ast::{
+    ast::{BigintLiteral, NumberLiteral},
+    AstKind,
+};
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::Error,
+};
+use oxc_macros::declare_oxc_lint;
+use oxc_span::Span;
+use regex::Regex;
+
+use crate::{context::LintContext, fixer::Fix, rule::Rule, AstNode};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("eslint-plugin-unicorn(numeric-separators-style): Enforce the style of numeric separators by correctly grouping digits")]
+#[diagnostic(severity(error), help("Invalid group length in numeric value."))]
+struct NumericSeparatorsStyleDiagnostic(#[label] pub Span);
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct NumericSeparatorsStyle {
+    only_if_contains_separator: bool,
+    hexadecimal: NumericBaseConfig,
+    binary: NumericBaseConfig,
+    octal: NumericBaseConfig,
+    number: NumericBaseConfig,
+}
+impl Default for NumericSeparatorsStyle {
+    fn default() -> Self {
+        Self {
+            only_if_contains_separator: false,
+            binary: NumericBaseConfig { group_length: 4, minimum_digits: 0 },
+            hexadecimal: NumericBaseConfig { group_length: 2, minimum_digits: 0 },
+            number: NumericBaseConfig { group_length: 3, minimum_digits: 5 },
+            octal: NumericBaseConfig { group_length: 4, minimum_digits: 0 },
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct NumericBaseConfig {
+    group_length: usize,
+    minimum_digits: usize,
+}
+impl NumericBaseConfig {
+    pub(self) fn set_numeric_base_from_config(&mut self, val: &serde_json::Value) {
+        if let Some(group_length) = val.get("groupLength").and_then(serde_json::Value::as_u64) {
+            self.group_length = usize::try_from(group_length).unwrap();
+        }
+        if let Some(minimum_digits) = val.get("minimumDigits").and_then(serde_json::Value::as_u64) {
+            self.minimum_digits = usize::try_from(minimum_digits).unwrap();
+        }
+    }
+}
+
+enum SeparatorDir {
+    Left,
+    Right,
+}
+
+fn add_separators(s: &mut String, dir: &SeparatorDir, config: &NumericBaseConfig) {
+    if s.len() < config.minimum_digits || s.len() < config.group_length + 1 {
+        return;
+    }
+
+    match dir {
+        SeparatorDir::Right => {
+            let mut pos = s.len();
+            while pos > config.group_length {
+                pos -= config.group_length;
+                s.insert(pos, '_');
+            }
+        }
+        SeparatorDir::Left => {
+            let mut pos = config.group_length;
+            while pos < s.len() {
+                s.insert(pos, '_');
+                pos += config.group_length + 1;
+            }
+        }
+    }
+}
```

Comment:
Shall we move all these below `impl Rule for NumericSeparatorsStyle {` so that `declare_oxc_lint` is at the top?